### PR TITLE
add a hidden debugging format

### PR DESCRIPTION
**Pull request by jonahwilliams**
Example:

```
RecordSize: Size(166.0, 202.0)
StoreGradient: id:0 Linear(from: (79.52133178710938, 170.67616271972656), to: (83.4103775024414, 174.56521606445312), colors: [Color(0x26000000),Color(0x02616161)], offsets: [0.20000000298023224, 0.8500000238418579], tileMode: 0
StoreGradient: id:1 Linear(from: (79.5, 142.8000030517578), to: (120.9000015258789, 142.8000030517578), colors: [Color(0x8c000000),Color(0x02616161)], offsets: [0.20000000298023224, 0.8500000238418579], tileMode: 0
StorePaint: id:0 Fill(Color(0xcc42a5f5), blendMode: 3, shader: null
StorePaint: id:1 Fill(Color(0xf542a5f5), blendMode: 3, shader: null
StorePaint: id:2 Fill(Color(0xff0d47a1), blendMode: 3, shader: null
StorePaint: id:3 Fill(Color(0xff42a5f5), blendMode: 3, shader: null
StorePaint: id:4 Fill(Color(0xffffffff), blendMode: 3, shader: 0
StorePaint: id:5 Fill(Color(0xffffffff), blendMode: 3, shader: 1
PathStart: id:0 Stroke
  moveTo(37.70000076293945, 128.89999389648438)
  lineTo(9.800000190734863, 101.0)
  lineTo(100.4000015258789, 10.399999618530273)
  lineTo(156.1999969482422, 10.399999618530273)
EndPath:
PathStart: id:1 Stroke
  moveTo(156.1999969482422, 94.0)
  lineTo(100.4000015258789, 94.0)
  lineTo(79.5, 114.9000015258789)
  lineTo(107.4000015258789, 142.8000030517578)
EndPath:
PathStart: id:2 Stroke
EndPath:
PathStart: id:3 Fill
  moveTo(79.5, 170.6999969482422)
  lineTo(100.4000015258789, 191.60000610351562)
  lineTo(156.1999969482422, 191.60000610351562)
  lineTo(156.1999969482422, 191.60000610351562)
  lineTo(107.4000015258789, 142.8000030517578)
EndPath:
PathStart: id:4 Fill
  moveTo(51.661590576171875, 142.81643676757812)
  lineTo(79.52133178710938, 114.9566879272461)
  lineTo(107.38107299804688, 142.81643676757812)
  lineTo(79.52133178710938, 170.67616271972656)
  close()
EndPath:
PathStart: id:5 Fill
  moveTo(79.52133178710938, 170.67616271972656)
  lineTo(107.38107299804688, 142.81643676757812)
  lineTo(111.2701187133789, 146.70547485351562)
  lineTo(83.4103775024414, 174.56521606445312)
  close()
EndPath:
PathStart: id:6 Fill
  moveTo(79.5, 170.6999969482422)
  lineTo(120.9000015258789, 156.39999389648438)
  lineTo(107.4000015258789, 142.8000030517578)
EndPath:
DrawPath: id:0 (0, null)
DrawPath: id:1 (0, null)
DrawPath: id:2 (1, null)
DrawPath: id:3 (2, null)
DrawPath: id:4 (3, null)
DrawPath: id:5 (4, null)
DrawPath: id:6 (5, null)

```

Useful for understanding how vector graphics will process a given SVG

### CODE REVIEW

**Comment by jonahwilliams** [comment]
This is from the Flutter Logo SVG

**Comment by jonahwilliams** [comment]
Path id 2 ends up empty, we should probably optimize this out and add some asserts around storing empty paths

File: packages/vector_graphics_compiler/lib/src/debug_format.dart [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  @override
+  void onDrawImage(int imageId, double x, double y, double width, double height,
+      Float64List? transform) {
+    buffer.writeln(
+        'DrawImage: id:$imageId (Rect.fromLTWH($x, $y, $width, $height), transform: $transform)');
+  }
+
+  @override
+  void onDrawPath(int pathId, int? paintId, int? patternId) {
+    buffer.writeln('DrawPath: id:$pathId ($paintId, $patternId)');
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/debug_format.dart:44]: nit: omit pattern id if null, and/or include a label for it?

File: packages/vector_graphics_compiler/lib/src/debug_format.dart [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    buffer.writeln(
+        'StoreImage: id:$imageId (format:$format, byteLength:${data.lengthInBytes}');
+  }
+
+  @override
+  void onLinearGradient(double fromX, double fromY, double toX, double toY,
+      Int32List colors, Float32List? offsets, int tileMode, int id) {
+    buffer.writeln(
+        'StoreGradient: id:$id Linear(from: ($fromX, $fromY), to: ($toX, $toY), '
+        'colors: [${colors.map(_intToColor).join(',')}], offsets: $offsets, tileMode: $tileMode');
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/debug_format.dart:68]: These would be easier to read if they're indented the way path commands are

File: packages/vector_graphics_compiler/lib/src/debug_format.dart [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      double? focalY,
+      Int32List colors,
+      Float32List? offsets,
+      Float64List? transform,
+      int tileMode,
+      int id) {
+    buffer.writeln(
+        'StoreGradient: id:$id Radial(center: ($centerX, $centerY), radius: $radius,'
+        ' focal: ($focalX, $focalY), colors: [${colors.map(_intToColor).join(',')}], offsets: $offsets, '
+        'transform: $transform, tileMode: $tileMode');
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/debug_format.dart:152]: same

File: packages/vector_graphics_compiler/lib/src/debug_format.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/debug_format.dart:68]: yeah good point, originally I was trying to keep everything on one line, but the identing is much nicer

**Review by dnfield — APPROVED** [review]
LGTM with nits

File: packages/vector_graphics_compiler/lib/src/debug_format.dart [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  @override
+  void onPathMoveTo(double x, double y) {
+    buffer.writeln('  moveTo($x, $y)');
+  }
+
+  @override
+  void onPathStart(int id, int fillType) {
+    buffer.writeln('PathStart: id:$id ${fillType == 0 ? 'Fill' : 'Stroke'}');
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/debug_format.dart:127]: fill type isn't the same as fill/stroke - 

```suggestion
    buffer.writeln('PathStart: id:$id ${fillType == 0 ? 'nonZero' : 'evenOdd'}');
```

Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/debug_format.dart:127]: 🦞

File: packages/vector_graphics_compiler/lib/src/debug_format.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/debug_format.dart]: Done

File: packages/vector_graphics_compiler/lib/src/debug_format.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/debug_format.dart]: Done

File: packages/vector_graphics_compiler/lib/src/debug_format.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/debug_format.dart]: Done

File: packages/vector_graphics_compiler/lib/src/debug_format.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/debug_format.dart]: Done

**Pull request merged**